packages/flutter_tools/lib/executable.dart

```diff
@@ -131,70 +131,78 @@ Future<Null> run(List<String> args, List<FlutterCommand> subCommands, {bool verb
     // Initialize the system locale.
     await intl.findSystemLocale();
 
-    return Chain.capture<Future<Null>>(() async {
+    Completer<Null> runCompleter = new Completer<Null>();
+    Chain.capture<Future<Null>>(() async {
       await runner.run(args);
       await _exit(0);
+      runCompleter.complete();
     }, onError: (dynamic error, Chain chain) {
-      if (error is UsageException) {
-        stderr.writeln(error.message);
-        stderr.writeln();
+      _handleToolError(error, chain, verbose, args).whenComplete(runCompleter.complete);
```

Comment:
Instead of `runCompleter.complete`, shouldn't it be `() => runCompleter.completeError(error, chain)`?  In tests, we're going to want to do the following (I have such a test in progress):

```
expect(() async {
  tools.run(...);
}, throwsProcessExit(0));
```